train.py

```diff
@@ -166,23 +185,131 @@ def update(engine, batch):
             input_ids, mc_token_ids, lm_labels, mc_labels, token_type_ids = batch
             logger.info(tokenizer.decode(input_ids[0, -1, :].tolist()))
             # if we dont send labels to model, it doesnt return losses
             lm_logits, mc_logits, *_ = model(
-                input_ids, token_type_ids=token_type_ids, mc_token_ids=mc_token_ids,
+                input_ids,
+                token_type_ids=token_type_ids,
+                mc_token_ids=mc_token_ids,
+            )
+            lm_logits_flat_shifted = (
+                lm_logits[..., :-1, :].contiguous().view(-1, lm_logits.size(-1))
             )
-            lm_logits_flat_shifted = lm_logits[..., :-1, :].contiguous().view(-1, lm_logits.size(-1))
             lm_labels_flat_shifted = lm_labels[..., 1:].contiguous().view(-1)
-            return (lm_logits_flat_shifted, mc_logits), (lm_labels_flat_shifted, mc_labels)
+            return (lm_logits_flat_shifted, mc_logits), (
+                lm_labels_flat_shifted,
+                mc_labels,
+            )
+
     evaluator = Engine(inference)
+    for name, metric in metrics.items():
+        metric.attach(evaluator, name)
+    if idist.get_rank() == 0:
+        pbar = common.ProgressBar(desc=f"Evaluation ({tag})", persist=False)
+        pbar.attach(evaluator)
+        evaluator.add_event_handler(
+            Events.COMPLETED,
+            lambda _: pbar.log_message(
+                "Validation: %s" % pformat(evaluator.state.metrics)
+            ),
+        )
+
+    return evaluator
+
+
+def train(local_rank, args):
+    rank = idist.get_rank()
+    logging.basicConfig(level=logging.INFO if local_rank in [-1, 0] else logging.WARN)
+    logger.warning(
+        "Running process %d", local_rank
+    )  # This is a logger.warning: it will be printed by all distributed processes
+    logger.info("Arguments: %s", pformat(args))
+
+    logger.info("Prepare tokenizer, pretrained model and optimizer.")
+    tokenizer_class = (
+        GPT2Tokenizer if "gpt2" in args.model_checkpoint else OpenAIGPTTokenizer
+    )  # cant use Autotokenizer because checkpoint could be a Path
+    tokenizer = tokenizer_class.from_pretrained(args.model_checkpoint)
+
+    model_class = (
+        GPT2DoubleHeadsModel
+        if "gpt2" in args.model_checkpoint
+        else OpenAIGPTDoubleHeadsModel
+    )
+    model = model_class.from_pretrained(args.model_checkpoint)
+    # Add special tokens if they are not already added
+    add_special_tokens_(model, tokenizer)
+
+    optimizer = AdamW(model.parameters(), lr=args.lr, correct_bias=True)
+
+    metrics = {
+        "Accuracy": Accuracy(output_transform=lambda x: (x[0][1], x[1][1])),
+        "nll": Loss(
+            torch.nn.CrossEntropyLoss(ignore_index=-100),
+            output_transform=lambda x: (x[0][0], x[1][0]),
+        ),
+    }
+    metrics.update(
+        {
+            "average_nll": RunningAverage(metrics["nll"]),
```

Comment:
Why do you use `RunningAverage` for validation ? Please, take a closer look at the original code, they `average_nll` to average the value accross distributed processes.